### PR TITLE
[REF] mail: activities,  revamp scss

### PR DESCRIPTION
**Pull request by stefanorigano**
Simplify/Reduce SCSS for the Activity component.
task-2731819.

requires:
- ~~https://github.com/odoo/odoo/pull/83260~~  -> Merged
- ~~https://github.com/odoo/odoo/pull/83495~~  -> Discarded 
- ~~https://github.com/odoo/odoo/pull/83498~~  -> Discarded 

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by stefanorigano** [comment]
Hello @seb-odoo , 
here a first PR, let's see if we're both on the same page :)

File: addons/mail/static/src/components/activity/activity.xml [hunk 1]
```diff
@@ -2,96 +2,96 @@
...
+                        </button>
                     </div>
 
                     <t t-if="activityView.areDetailsVisible">
                         <div class="o_Activity_details">
-                            <dl class="dl-horizontal">
-                                <t t-if="activityView.activity.type">
-                                    <dt>Activity type</dt>
-                                    <dd class="o_Activity_type">
+                            <table class="table table-sm mt-2 mb-3">
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.xml:56]: Isn't it conceptually wrong to position elements with a table?
Should probably use either flex layout with some fixed-width elements, or grid layout, depending on the desired effect?

File: addons/mail/static/src/components/activity/activity.scss [hunk 2]
```diff
@@ -2,68 +2,35 @@
...
-
 .o_Activity_detailsUserAvatar {
-    margin-inline-end: map-get($spacers, 2);
-    object-fit: cover;
-    height: 18px;
-    width: 18px;
+    height: 1.4em;
+    width: 1.4em;
+    vertical-align: text-bottom;
+    margin-inline-end: map-get($spacers, 1);
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.scss:9]: Why is this one not becoming a `mr-1` for example?

**Review by seb-odoo — COMMENTED** [review]
Hello, here are a few comments in code as a first check.
And also: 
- runbot is red, so it's hard to see the result
- the PR has conflicts
- it would be easier to review to split the changes that are "just" moving css rules into classes into a different PR (without changing style) and making a separate PR for the style changes

File: addons/mail/static/src/components/activity/activity.scss [hunk 3]
```diff
@@ -2,68 +2,35 @@
...
-    justify-content: center;
-}
-
-.o_Activity_toolButton {
-    padding-top: map-get($spacers, 0);
-}
-
-.o_Activity_tools {
-    display: flex;
+.o_Activity_sidebar, .o_Activity_user {
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.scss:24]: Also avoid `,` rules as they get quite hard to follow sometimes.

All the layout for `o_Activity_user` should only be in one place, and same for `o_Activity_sidebar`.

If they need common style, we should either:
- create a new class, to apply on both
- define a mixin and use it on both

File: addons/mail/static/src/scss/variables/primary_variables.scss [hunk 4]
```diff
@@ -1,4 +1,4 @@
-$o-mail-thread-avatar-size: 36px !default;
+$o-mail-thread-avatar-size: $o-avatar-size * .4 !default;
```
Comment (seb-odoo) [inline addons/mail/static/src/scss/variables/primary_variables.scss:1]: Is there a specific reason or do we have a guideline to omit the 0 before decimal? I find it harder to read, it looks a bit like a typo and at glance when scanning code it's easier to miss the lonely dot and think it's 4.

File: addons/mail/static/src/components/activity/activity.xml (hunk 1 above)
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.xml:56]: mmm... seems the perfect case for a <table>, we are presenting data.

File: addons/mail/static/src/components/activity/activity.scss (hunk 2 above)
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.scss:9]: For rtl languages since the avatar is rendered as inline text

File: addons/mail/static/src/components/activity/activity.scss [hunk 5]
```diff
@@ -2,68 +2,35 @@
...
-    padding: map-get($spacers, 2);
-}
-
 .o_Activity_detailsUserAvatar {
-    margin-inline-end: map-get($spacers, 2);
-    object-fit: cover;
-    height: 18px;
-    width: 18px;
+    height: 1.4em;
+    width: 1.4em;
```
[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.scss:7]: agreed 👍

File: addons/mail/static/src/components/activity/activity.scss [hunk 6]
```diff
@@ -79,108 +46,22 @@
...
 
 .o_Activity_toolButton {
-    opacity: 0.5;
-    color: gray('500');
+    @include o-hover-text-color();
 
-    &:hover {
-        opacity: 1;
-        color: gray('600');
+    .o_Activity:not(:hover) & {
```
[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.scss:60]: > isn't "not hover" just standard style?

not-hover is the "exception" that use the rule `opacity: .6;`, since by default is `opacity: 1;`

Otherwise I'll have to define the rule twice, eg:
```scss
.o_Activity & {
        opacity: .6;
 }
.o_Activity:hover & {
        opacity: 1;
 }
```

File: addons/mail/static/src/components/activity/activity.scss [hunk 7]
```diff
@@ -79,108 +46,22 @@
...
     }
 }
 
-.o_Activity_userAvatar {
-    @extend .rounded-circle;
-}
-
-.o_Activity_userName {
-    color: gray('500');
+.o_Activity_details tr:first-child() > * {
```
[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.scss:65]: Yes, thanks.
It's the (the only?) reason why runbot is red indeed ;)

File: addons/mail/static/src/components/activity/activity.scss (hunk 3 above)
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.scss:24]: Ok, I'll reuse the same rule for both selectors. 

> If they need common style, we should either:
>- create a new class, to apply on both elements in the DOM
>- define a mixin and use it on both in css

yep, pseudo-classes are very convenient too :)

File: addons/mail/static/src/components/activity/activity.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/components/activity/activity.scss:60]: Well except `o-hover-text-color` also puts the style on `&:hover, &:focus, &.focus` and you don't negate the others, so it's not working as intended.

It's much easier to understand when defining both rules like the existing code (just change value from 0.5 to 0.6 if you need) and like in your example.

Also what is this extra `&` exactly? Same comment here as elsewhere, don't use specificity and put class where they are needed.

File: addons/mail/static/src/components/activity/activity.xml (hunk 1 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/activity/activity.xml:56]: But a table only makes sense for tabular data, it's not tabular data if the value in each column has a different meaning on each row.

There is no "column header" that would make sense for all rows, or vice versa.

Another hint that it's probably not good as a table is that there is no "for loop" to generate the rows in a similar pattern.

File: addons/mail/static/src/components/activity/activity.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.scss:60]: > Well except o-hover-text-color also puts the style on &:hover, &:focus, &.focus and you don't negate the others, so it's not working as intended.

I think you missed the point. We don't care if `o_Activity` is :focus, we only care when the user interact hovering it ;)

> It's much easier to understand when defining both rules like the existing code

Yes but it's 3 lines more + an unnecessary override. I prefer to add 3 lines of notes if it's not clear.

> Also what is this extra & exactly?

`&` is not "extra" ,  it's the sass parent selector.
https://sass-lang.com/documentation/style-rules/parent-selector

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/components/activity/activity.scss:61]: We definitely don't want to ever use parent selector, it's just magically adding style on elements from random context.
Same as usual, if we need a style on the parent in that case, then the parent should have a class for that style on itself when necessary.

[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.scss:61]: > We definitely don't want to ever use parent selector, it's just magically adding style on elements from random context.

You do realise that it's equivalent to write  `.o_Activity:not(:hover) .o_Activity_toolButton {}`, right ?
There is nothing "magic"... it's just taking advantage of the preprocessor for what it is supposed to do, simplify the code.

Anyways... it's my understanding we could spend ages debating on a single line, but we don't have the time ^^
I'll try to stay as close as possible to what you want.

Could you just share with me the guidelines your team uses?
What you suggest is not in the official ones  https://github.com/odoo/odoo/wiki/SCSS-coding-guidelines

[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/components/activity/activity.scss:61]: > > We definitely don't want to ever use parent selector, it's just magically adding style on elements from random context.
> 
> You do realise that it's equivalent to write `.o_Activity:not(:hover) .o_Activity_toolButton {}`, right ? There is nothing "magic"... it's just taking advantage of the preprocessor for what it is supposed to do, simplify the code.

The underlying problem with this rule is that it deals with higher CSS specificity than `.o_Activity_toolButton` local rules. At first this doesn't look like a big deal, but actually that's precisely what makes most of the SCSS a nightmare to maintain when CSS specificity is spread all over the style of the whole application.

The more maintainable approach consists of this: all style-rules target immediately the classname of elements to style, and contextual styling come only from "modifier" classnames on these elements. That way, CSS specificity is always the same for all static styles (= 1) and for all contextual styles (= 2). This design incites all styling on an element being pretty local in the code, which is very good for code maintenance.

Unfortunately, doing so is quite cumbersome right now, as you need to implement modifier classnames by hand and integrate some built-in feature of CSS in JS/XML like `:hover`... We actually did this in many places and we are quite happy to have them instead of dealing with specificity in CSS everywhere. So it's easier to maintain, but unfortunately it requires more effort to make it work initially.

TBH, it's not urgent to do the proper thing right now. At least, the problem with `.o_Activity:not(:hover) .o_Activity_toolButton` is only limited to elements from the `Activity` component, which is better than cross-component styling for instance! But this clearly worse than the approach with modifier classnames, because potentially anything in the SCSS file can have styling effect on any part of the component rather than just the style on element.

Let's say discuss code is around 80 SCSS files and each average 100 LOCs, and all styles that target a given classname average to about 10 LOCs:
1) Dealing with different CSS specificities in all discuss SCSS files, I would need to know all combination of CSS styling of all files at all time, which is 80 * 100 LOCs, so **8000 LOCs**.
2) Dealing with different CSS specificities in each discuss SCSS file, I would need to know all combination of CSS styling of single file at a given time, so **100 LOCs**
3) The approach of minimising CSS specificity wth immediate target classname and modifier classnames, so **10 LOCs**.

Just quantitatively, 8000 > 100 > 10.
Please note that dealing with CSS specificity in cases 1 and 2 make branching for evaluating styles to be applied on some elements. These are much much harder to work with. IMO I'd say the effort to maintain CSS is exponential so even between 100 > 10 the maintenance cost is quite significant.

Thank you for attending my TED Talk 😅

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/components/activity/activity.scss:61]: To complement what Alex just posted.

We follow more or less BEM guidelines (Block Element Modifier) slightly adapted to how things are usually named in Odoo (often to simplify a bit). I think it would be interesting that you read on the subject if you haven't heard of it before.

Each rule should only target one specific class, not children, not parent, not siblings, and there should be no nested classes either.

The goal is when we look at any DOM element, we want to know exactly which style applies to it by looking only at its own class attribute.
Because not doing it this way means any line of CSS anywhere in the code could have an impact on any possible element, which is completely unmaintainable, like existing CSS files.


In this particular case, if we need to apply style on `.o_Activity_toolButton` conditionally, then the element with `.o_Activity_toolButton` should receive the conditional class (example: `.o-activityFocused`) with a `t-aff` or `t-attf` when it is necessary, which means we need to store the focus state in our modelization so that it gets rendered correctly with the class on the element from the template. Of course I'm available to help you with that part if you need, I can give you the code necessary to add the class in the DOM, so that you can use it for style.

[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/components/activity/activity.scss:61]: Very happy to see that we share a lot of opinions.
I don't think it's a surprise: this shows we are able to take a step back with the tools we use and pinpoint precisely some of their issues.

> Modularity & Reusability

Fully agree that currently the code of Discuss component is not reusable, which is clearly not efficient.

Actually all of this comes from our initial experience with Odoo Widgets / OWL components: code architecture with them strongly motivates to design the code around UI elements in DOM, and not really care with the state of application. For very small and simple codebase, this is decent while giving fast results. But when your modelling becomes complex due to functional scope, this scales very poorly.

In Discuss, we work with models first and then components. Components is about UI so if 2 parts of UI share some styles and it's not a coincidence, this means these UI elements share a same model. Our priority is fixing our mistakes with Widget/Component of the past and properly use a model-oriented architecture. Reusability of component is a much lower priority at the moment, as things are still relatively maintainable even with code duplication.
Also the huge work on model-oriented development will fix component reusability, because we will have proper model like `EntriesList`, so we will define components are proper UI things rather than very specific components.

So I fully agree with you on that part, and it takes a lot of time to improve that properly.

> Combined selectors

If the 2 styles are related and it's not by coincidence, this means there is an underlying concept that is shared.
In our model-oriented way of developing the Discuss codebase, there is a model that is shared between the 2 elements.

Instead of guessing what is the concept, prefer to actually introduce it. If you do so, then you don't actually need combined selectors.

> Split selectors in layout/style

We are not sure to keep this split. We just followed the idea is some other SCSS files and like some advantages of this split ("layout" is macro-view of style where  changes can affect the whole screen, while "style" is very local and safe to change).

It's a bit cumbersome to make the split, even though there's clearly a gain in the split.
IMO this is not a big deal as the split layout/style is arbitrary split in a single SCSS file, and if we follow the methodology of minimising CSS specificity with BEM, then it's a non-issue w.r.t code maintainability.

Just my opinion, but I think layout/style split makes sense, but unfortunately everything is mixed in one technology (CSS), and this declarative language a bit of a mess...

File: addons/mail/static/src/components/activity/activity.xml (hunk 1 above)
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.xml]: Adapted avoiding to use the `<table>` tag 👍 
The new implementation uses generic `<div>`s in combination with default bts classes.

Please note that this is a fix, not a design improvement. Indeed currently we use  `dl-horizontal` that's discontinued , causing the layout to brake.

Comment (seb-odoo) [reply addons/mail/static/src/components/activity/activity.xml]: If possible it would be better to merge the fix in a separate PR.
As I see it, this PR could be split in 3 different PRs that would be much easier to review and merge faster separately:

- fix/adaptation for dl
- in other places, use classes instead of css (without design change)
- design changes (only after everything else is clean)

Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.xml]: ok, thanks !

> design changes (only after everything else is clean)

There are not real design changes, just minor differences caused by the use of default classes.
I'll open a separate PR for the  dl 👍

File: addons/web/static/src/legacy/scss/ui.scss [hunk 8]
```diff
@@ -95,6 +95,26 @@ span.o_force_ltr {
...
+    vertical-align: text-bottom;
+}
+
+// == Thread avatar image
+//------------------------------------------------------------------------------
+.o_avatar_thread {
+    width: $o-avatar-thread-size;
+    height: $o-avatar-thread-size;
+    object-fit: cover;
+}
```
Comment (seb-odoo) [inline addons/web/static/src/legacy/scss/ui.scss:116]: The concept of "thread" does not exist in the `web` module. This should either be defined in `mail`, or find another name.

Comment (stefanorigano) [reply addons/web/static/src/legacy/scss/ui.scss:116]: The `thread` name and  the `web` locations seemed ok to me  because:
- the class is used by more than one module (`mail`and `im_livechat`)
- both occurrences are in "threads" content

Any advice?

Comment (seb-odoo) [reply addons/web/static/src/legacy/scss/ui.scss:116]: There is a dependency on `mail` for all modules that use "thread, including `im_livechat`, so the move to `mail` would be ok in that regard.

Comment (stefanorigano) [reply addons/web/static/src/legacy/scss/ui.scss:116]: Ok, I'll move it to `mail` and I'll use the  `o_mail_ thread_avatar ` 👍

**Comment by stefanorigano** [comment]
@seb-odoo , I preferred to move forward and proceed with SCSS optimisations only for now.
Dependencies with now discarded PRs have been removed.

Ready for your review

File: addons/mail/static/src/components/activity/activity.xml [hunk 9]
```diff
@@ -64,8 +67,8 @@
                                 <div t-if="activityView.activity.creator" class="d-md-table-row mb-3">
                                     <div class="d-md-table-cell font-weight-bold text-md-right m-0 py-md-1 px-md-4">Created</div>
                                     <div class="o_Activity_detailsCreation d-md-table-cell py-md-1 pr-4">
-                                        <t t-esc="activityView.formattedCreateDatetime"/>, by
-                                        <img class="o_Activity_detailsUserAvatar o_Activity_detailsCreatorAvatar" t-attf-src="/web/image/res.users/{{ activityView.activity.creator.id }}/avatar_128" t-att-title="activityView.activity.creator.nameOrDisplayName" t-att-alt="activityView.activity.creator.nameOrDisplayName"/>
+                                        <t t-esc="activityView.formattedCreateDatetime"/>, <br class="d-md-none"/>by
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.xml:70]: Don't use `d-none`, prefer not rendering useless elements: `<br class="d-md-none"/>` -> `<br t-if="messaging.device.isMobile"/>`

File: addons/mail/static/src/components/activity/activity.xml [hunk 10]
```diff
@@ -2,52 +2,55 @@
...
+                                'text-success': activityView.activity.state === 'planned',
+                                'text-warning': activityView.activity.state === 'today',
                             }"
                         >
-                            <t t-esc="activityView.delayLabel"/>
+                            <b t-esc="activityView.delayLabel"/>
                         </div>
                         <t t-if="activityView.activity.summary">
-                            <div class="o_Activity_summary">
+                            <b class="o_Activity_summary me-2 h6">
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.xml:35]: As I posted on discord, these `h6` are making the bottom line wrongly aligned on the avatar.

File: addons/mail/static/src/components/activity/activity.xml [hunk 11]
```diff
@@ -2,52 +2,55 @@
 <templates xml:space="preserve">
 
     <t t-name="mail.Activity" owl="1">
-        <div class="o_Activity" t-attf-class="{{ className }}" t-on-click="activityView.onClickActivity" t-ref="root">
+        <div class="o_Activity d-flex p-2" t-attf-class="{{ className }}" t-on-click="activityView.onClickActivity" t-ref="root">
             <t t-if="activityView">
-                <div class="o_Activity_sidebar">
-                    <div class="o_Activity_user">
+                <div class="o_Activity_sidebar mr-3">
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.xml:7]: Ok if we handle it on another PR, but I feel it would look better if we increased the left spacing on the avatar for message, rather than reducing it for activities.

File: addons/mail/static/src/components/activity/activity.xml [hunk 12]
```diff
@@ -2,52 +2,55 @@
...
                             <div class="o_Activity_userName">
                                 <t t-esc="activityView.assignedUserText"/>
                             </div>
                         </t>
-                        <a class="o_Activity_detailsButton btn btn-link" t-on-click="activityView.onClickDetailsButton" role="button">
+                        <a
+                            class="o_Activity_detailsButton btn py-0"
+                            t-att-class="activityView.areDetailsVisible ? 'text-primary' : 'btn-link btn-primary'"
+                            t-on-click="activityView.onClickDetailsButton"
+                            role="button">
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.xml:53]: Closing tag `>` on a new line, with same indent level as the opening tag.

File: addons/mail/static/src/components/activity/activity.xml [hunk 13]
```diff
@@ -116,10 +118,10 @@
                     </t>
 
                     <t t-if="activityView.activity.canWrite">
-                        <div name="tools" class="o_Activity_tools">
+                        <div name="tools" class="o_Activity_tools d-flex">
                             <t t-if="!activityView.fileUploader">
                                 <Popover position="'right'" title="activityView.markDoneText">
-                                    <button class="o_Activity_toolButton o_Activity_markDoneButton btn btn-link" t-att-title="activityView.markDoneText">
+                                    <button class="o_Activity_toolButton o_Activity_markDoneButton btn btn-link btn-primary pt-0 pl-0" t-att-title="activityView.markDoneText">
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.xml:124]: Aren't `btn-link` and `btn-primary` "conflicting" style classes?

File: addons/mail/static/src/components/activity/activity.xml [hunk 14]
```diff
@@ -128,14 +130,14 @@
                                 </Popover>
                             </t>
                             <t t-if="activityView.fileUploader">
-                                <button class="o_Activity_toolButton o_Activity_uploadButton btn btn-link" t-on-click="activityView.onClickUploadDocument">
+                                <button class="o_Activity_toolButton o_Activity_uploadButton btn btn-link btn-primary pt-0 pl-0" t-on-click="activityView.onClickUploadDocument">
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.xml:133]: same remark for all others

File: addons/mail/static/src/components/activity/activity.xml (hunk 13 above)
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.xml:124]: > Aren't btn-link and btn-primary "conflicting" style classes?

Nope, we handle this combination https://github.com/odoo/odoo/blob/master/addons/web/static/src/webclient/webclient.scss#L143

Note: we'll review the way how we handle this kind of buttons, but it will be after the migration.

File: addons/mail/static/src/components/activity/activity.xml (hunk 11 above)
Comment (stefanorigano) [reply addons/mail/static/src/components/activity/activity.xml:7]: ok. For reference:
> Add ml-1 and flex-shrink-0 to  o_Activity_sidebar.

File: addons/mail/static/src/components/activity/activity.xml [hunk 15]
```diff
@@ -128,14 +131,14 @@
...
                                     <i class="fa fa-upload"/> Upload Document
                                 </button>
                             </t>
-                            <button class="o_Activity_toolButton o_Activity_editButton btn btn-link" t-on-click="activityView.onClickEdit">
+                            <button class="o_Activity_toolButton o_Activity_editButton btn btn-link btn-primary pt-0" t-on-click="activityView.onClickEdit">
                                 <i class="fa fa-pencil"/> Edit
                             </button>
-                            <button class="o_Activity_toolButton o_Activity_cancelButton btn btn-link" t-on-click="activityView.onClickCancel" >
+                            <button class="o_Activity_toolButton o_Activity_cancelButton btn btn-link btn-primary pt-0" t-on-click="activityView.onClickCancel" >
```
Comment (seb-odoo) [inline addons/mail/static/src/components/activity/activity.xml:141]: I really dislike changing padding on buttons because it changes the expected and natural click area.
I'll let this one pass if a redesign is planned but I still wanted to mention it.

Comment (seb-odoo) [reply addons/mail/static/src/components/activity/activity.xml:141]: Also note that it would be nice if there was some margin between the buttons to avoid the clicking area to touch each other.

**Review by seb-odoo — APPROVED** [review]
@robodoo r+ rebase-ff
